File: Codellama-34b-Instruct/Template-and-context-final/serial_construct_self_clause.f90
fortran
#ifndef T1
!T1:serial construct self clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        INTEGER :: device_type
        LOGICAL :: self_clause_present
        
        ! Check if the self clause is present
        self_clause_present = .FALSE.
        IF (PRESENT(self_clause)) THEN
          self_clause_present = .TRUE.
        END IF
        
        ! Check if the device type is the same as the current device
        device_type = acc_get_device_type()
        IF (device_type .EQ. acc_device_host) THEN
          IF (self_clause_present) THEN
            errors = errors + 1
          END IF
        ELSE IF (device_type .EQ. acc_device_not_host) THEN
          IF (.NOT. self_clause_present) THEN
            errors = errors + 1
          END IF
        END IF
        
        IF (errors .EQ. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END
#endif

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
#ifndef T1
        LOGICAL :: test1
#endif
        failed = .FALSE.
        failcode = 0
#ifndef T1
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
#endif
        CALL EXIT (failcode)
      END PROGRAM

File: Codellama-34b-Instruct/Template-and-context-final/serial_construct_self_clause.cpp
#include <iostream>
#include <cstdlib>
#include <cmath>
#include <acc/acc.h>

#define NUM_TEST_CALLS 100
#define SEED 12345

int test1() {
    int err = 0;
    srand(SEED);

    // Create a serial construct with a self clause
    acc_serial_construct(acc_self_clause);

    // Execute some code in the serial construct
    for (int i = 0; i < 10; i++) {
        std::cout << "Hello from the serial construct!" << std::endl;
    }

    // Check if the self clause was executed on the local device
    if (acc_self_clause) {
        std::cout << "The self clause was executed on the local device." << std::endl;
    } else {
        std::cout << "The self clause was not executed on the local device." << std::endl;
    }

    // Check if the self clause was executed on the current device
    if (acc_self_clause) {
        std::cout << "The self clause was executed on the current device." << std::endl;
    } else {
        std::cout << "The self clause was not executed on the current device." << std::endl;
    }

    return err;
}

int main() {
    int failcode = 0;
    int failed;

    // Run the test 100 times
    for (int x = 0; x < NUM_TEST_CALLS; x++) {
        failed = test1();
        if (failed != 0) {
            failcode = failcode + (1 << x);
        }
    }

    return failcode;
}